hpx/lcos/local/latch.hpp

```diff
@@ -0,0 +1,161 @@
+//  Copyright (c) 2015 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+/// \file hpx/lcos/local/latch.hpp
+
+#if !defined(HPX_LCOS_LATCH_APR_18_2015_0925PM)
+#define HPX_LCOS_LATCH_APR_18_2015_0925PM
+
+#include <hpx/lcos/local/detail/condition_variable.hpp>
+#include <hpx/lcos/local/spinlock.hpp>
+
+///////////////////////////////////////////////////////////////////////////////
+namespace hpx { namespace lcos { namespace local
+{
+    /// Latches are a thread coordination mechanism that allow one or more
+    /// threads to block until an operation is completed. An individual latch
+    /// is a singleuse object; once the operation has been completed, the latch
+    /// cannot be reused.
+    ///
+    /// A latch maintains an internal counter_ that is initialized when the
+    /// latch is created. Threads may block at a synchronization point waiting
+    /// for counter_ to be decremented to 0. When counter_ reaches 0, all such
+    /// blocked threads are released.
+    ///
+    /// Calls to countdown_and_wait() , count_down() , wait() , and is_ready()
+    /// behave as atomic operations.
+    ///
+    /// \note   A \a local::latch is not a LCO in the sense that it has no
+    ///         global id and it can't be triggered using the action (parcel)
+    ///         mechanism. Use lcos::latch instead if this is required.
+    ///         It is just a low level synchronization primitive allowing to
+    ///         synchronize a given number of \a threads.
+    class latch
+    {
+#if defined(BOOST_NO_CXX11_DELETED_FUNCTIONS)
+    private:
+        latch (latch const&);
+        latch (latch&&);
+
+        latch& operator= (latch const&);
+        latch& operator= (latch&&);
+#else
+    public:
+        latch (latch const&) = delete;
+        latch (latch&&) = delete;
+
+        latch& operator= (latch const&) = delete;
+        latch& operator= (latch&&) = delete;
+#endif
+
+    private:
+        typedef lcos::local::spinlock mutex_type;
+
+    public:
+        /// Initialize the latch
+        ///
+        /// Requires: count >= 0.
+        /// Synchronization: None
+        /// Postconditions: counter_ == count.
+        ///
+        explicit latch(std::ptrdiff_t count)
+          : counter_(count)
+        {}
+
+        /// Requires: No threads are blocked at the synchronization point.
+        ///
+        /// \note May be called even if some threads have not yet returned
+        ///       from wait() or count_down_and_wait(), provided that counter_
+        ///       is 0.
+        /// \note The destructor might not return until all threads have exited
+        ///       wait() or count_down_and_wait().
```

Comment:
I think this note does not reflect the implementation. In addition it is contradicting the documented requirement that no other threads are blocked in a wait.